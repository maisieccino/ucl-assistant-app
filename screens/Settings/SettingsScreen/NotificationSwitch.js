import PropTypes from "prop-types"
import React, { Component } from "react"
import { ActivityIndicator, Alert, Switch } from "react-native"
import { connect } from "react-redux"

// import { registerForNotifications } from "../../../actions/notificationsActions"
import { Horizontal } from "../../../components/Containers"
import { BodyText, ErrorText } from "../../../components/Typography"
import common from "../../../styles/common"

class NotificationSwitch extends Component {
  static mapStateToProps = (state) => ({
    changing: state.notifications.stateChanging,
    error: state.notifications.stateChangeError,
    registered: state.notifications.registered,
    token: state.user.token,
  })

  static mapDispatchToProps = (dispatch) => ({
    register: (token) => dispatch(token/* registerForNotifications(token) */),
  })

  static propTypes = {
    changing: PropTypes.bool,
    error: PropTypes.string,
    register: PropTypes.func,
    registered: PropTypes.bool,
    token: PropTypes.string,
  }

  static defaultProps = {
    changing: false,
    error: ``,
    register: () => { },
    registered: false,
    token: ``,
  }

  // state = {
  //   spin: false,
  // }

  async onSwitchChange() {
    const { registered, register, token } = this.props
    if (!registered) {
      await this.setState({ spin: true })
      Alert.alert(
        `Register for notifications?`,
        `The only information stored on our servers is a unique key used to send notifications to your device. Are you sure you want to continue?`,
        [
          {
            onPress: () => this.setState({ spin: false }),
            style: `cancel`,
            text: `Cancel`,
          },
          {
            onPress: () => {
              this.setState({ spin: false })
              register(token)
            },
            text: `Continue`,
          },
        ],
      )
    } else {
      // not yet implemented on API
      this.setState({ spin: false })
      Alert.alert(
        `Can't yet unregister notifications.`,
        `Can't yet unregister for notifications.`,
      )
    }
  }

  render() {
    const { registered, changing, error } = this.props
    const { spin } = this.state
    return (
      <>
        {error.length > 0 && (
          <ErrorText>
            Error:
            {error}
          </ErrorText>
        )}
        <Horizontal>
          <BodyText style={common.flex}>Enable notifications</BodyText>
          {changing || spin ? (
            <ActivityIndicator />
          ) : (
            <Switch
              onValueChange={(b) => this.onSwitchChange(b)}
              value={registered}
            />
          )}
        </Horizontal>
      </>
    )
  }
}

export default connect(
  NotificationSwitch.mapStateToProps,
  NotificationSwitch.mapDispatchToProps,
)(NotificationSwitch)
