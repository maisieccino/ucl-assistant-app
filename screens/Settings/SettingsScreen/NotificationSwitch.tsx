import React from "react"
import { ActivityIndicator, Alert, Switch } from "react-native"
import { connect, ConnectedProps } from "react-redux"

/* import {
  registerForNotifications
} from "../../../actions/notificationsActions" */
import { Horizontal } from "../../../components/Containers"
import { BodyText, ErrorText } from "../../../components/Typography"
import { AppStateType } from "../../../configureStore"
import common from "../../../styles/common"

interface State {
  spin: boolean,
}

class NotificationSwitch extends React.Component<PropsFromRedux, State> {
  constructor(props) {
    super(props)
    this.state = {
      spin: false,
    }
  }

  onSwitchChange = async () => {
    const { registered, register, token } = this.props
    if (!registered) {
      await this.setState({ spin: true })
      Alert.alert(
        `Register for notifications?`,
        `The only information stored on our servers is a unique key`
        + ` used to send notifications to your device.`
        + ` Are you sure you want to continue?`,
        [
          {
            onPress: () => this.setState({ spin: false }),
            style: `cancel`,
            text: `Cancel`,
          },
          {
            onPress: () => {
              this.setState({ spin: false })
              register(token)
            },
            text: `Continue`,
          },
        ],
      )
    } else {
      // not yet implemented on API
      this.setState({ spin: false })
      Alert.alert(
        `Can't yet unregister notifications.`,
        `Can't yet unregister for notifications.`,
      )
    }
  }

  render() {
    const { registered, changing, error } = this.props
    const { spin } = this.state
    return (
      <>
        {error.length > 0 && (
          <ErrorText>
            Error:
            {error}
          </ErrorText>
        )}
        <Horizontal>
          <BodyText style={common.flex}>Enable notifications</BodyText>
          {changing || spin ? (
            <ActivityIndicator />
          ) : (
              <Switch
                onValueChange={this.onSwitchChange}
                value={registered}
              />
          )}
        </Horizontal>
      </>
    )
  }
}

const connector = connect(
  (state: AppStateType) => ({
    changing: state.notifications.stateChanging,
    error: state.notifications.stateChangeError,
    registered: state.notifications.registered,
    token: state.user.token,
  }),
  (dispatch) => ({
    register: (token) => dispatch(token/* registerForNotifications(token) */),
  }),
)

type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(NotificationSwitch)
