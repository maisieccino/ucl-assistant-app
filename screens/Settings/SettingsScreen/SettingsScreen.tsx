import type { BottomTabNavigationProp } from "@react-navigation/bottom-tabs"
import {
  CommonActions,
  CompositeNavigationProp,
} from "@react-navigation/native"
import type { StackNavigationProp } from "@react-navigation/stack"
import Constants from "expo-constants"
import * as IntentLauncherAndroid from "expo-intent-launcher"
import * as StoreReview from 'expo-store-review'
import React from "react"
import {
  Alert,
  Linking,
  Platform,
  StyleSheet,
  View,
} from "react-native"
import CheckBox from 'react-native-check-box'
import { connect, ConnectedProps } from "react-redux"

import {
  setShouldTrackAnalytics as setShouldTrackAnalyticsAction,
  signOut as signOutAction,
  UserDispatch,
} from "../../../actions/userActions"
import { SmallButton } from "../../../components/Button"
import { Horizontal, Page } from "../../../components/Containers"
import TextInput from "../../../components/Input/TextInput"
// import NotificationSwitch from "./NotificationSwitch"
import LiveIndicator from "../../../components/LiveIndicator"
import {
  BodyText,
  HeaderText,
  Link,
} from "../../../components/Typography"
import { AppStateType } from "../../../configureStore"
import { AnalyticsManager, ClipboardManager, MailManager } from "../../../lib"
import type { MainTabNavigatorParamList } from "../../../navigation/MainTabNavigator"
import type { RootStackParamList } from "../../../navigation/RootNavigation"
import * as packageJson from '../../../package.json'
import common from "../../../styles/common"
import type { SettingsNavigatorParamList } from "../SettingsNavigator"

const {
  repository: {
    url: githubURL,
  },
  version,
} = packageJson

const styles = StyleSheet.create({
  faqButton: {
    marginBottom: 10,
  },
  feedbackButton: {
    alignSelf: `flex-start`,
    marginVertical: 10,
  },
  // notificationSettingsButton: {
  //   marginTop: 10,
  // },
  releaseChannel: {
    alignItems: `center`,
    flexDirection: `row`,
    justifyContent: `flex-start`,
    marginBottom: 10,
  },
  section: {
    marginBottom: 15,
    marginTop: 15,
  },
  settingView: {
    flex: 1,
    flexDirection: `row`,
  },
  settingsCheckBox: {
    flex: 0.1,
    justifyContent: `center`,
  },
  settingsText: {
    flex: 0.9,
  },
  signOut: {
    marginBottom: 10,
    marginTop: 10,
  },
  textWithUpperMargin: {
    marginTop: 10,
  },
})

interface Props extends PropsFromRedux {
  navigation: CompositeNavigationProp<
    StackNavigationProp<SettingsNavigatorParamList>,
    CompositeNavigationProp<
      BottomTabNavigationProp<MainTabNavigatorParamList>,
      StackNavigationProp<RootStackParamList>
    >
  >,
}

interface State {
  isSigningOut: boolean,
}

export class SettingsScreen extends React.Component<Props, State> {
  static navigationOptions = {
    headerShown: false,
  }

  constructor(props: Props) {
    super(props)
    this.state = {
      isSigningOut: false,
    }
  }

  componentDidUpdate(_, prevState: State): void {
    const { user, navigation } = this.props
    if (prevState.isSigningOut && user.token === ``) {
      const action = CommonActions.reset({
        index: 0,
        routes: [{ name: `Splash` }],
      })
      navigation.dispatch(action)
    }
  }

  launchNotificationSettings = (): void => {
    // note that this will only work on standalone apps
    // because the bundleIdentifier/packageName will
    // only be used when an APK/IPA is built
    if (Platform.OS === `android`) {
      IntentLauncherAndroid.startActivityAsync(
        IntentLauncherAndroid.ACTION_APP_NOTIFICATION_SETTINGS,
        {
          "android.provider.extra.APP_PACKAGE":
            Constants.manifest.android.package,
        } as any,
      )
    } else {
      // is iOS
      Linking.openURL(
        `app-settings://notification/${
          Constants.manifest.ios.bundleIdentifier
        }`,
      )
    }
  }

  toggleAnalytics = (): void => {
    const {
      user: {
        settings: { shouldTrackAnalytics: value },
      },
      setShouldTrackAnalytics,
    } = this.props

    setShouldTrackAnalytics(!value)
  }

  signOut = (): void => {
    const { signOut } = this.props
    signOut()
    this.setState({ isSigningOut: true })
  }

  navigateToFAQs = (): void => {
    const { navigation } = this.props
    navigation.navigate(`FAQ`)
    AnalyticsManager.logEvent(AnalyticsManager.events.SETTINGS_VIEW_FAQS)
  }

  copyTokenToClipboard = async (): Promise<void> => {
    const { user: { token } } = this.props
    await ClipboardManager.setString(token)
    Alert.alert(`Copied`, `Token copied to clipboard.`)
  }

  giveFeedback = (): void => {
    const { user: { upi } } = this.props
    const { deviceName, platform, manifest: { releaseChannel } } = Constants
    MailManager.composeAsync({
      body: `I've been using UCL Assistant and I just wanted `
        + `to tell you ... \n\n`
        + `Technical Information\n${JSON.stringify({
          deviceName,
          platform,
          releaseChannel,
          upi,
        }, null, 2)}`,
      recipients: [`isd.apiteam@ucl.ac.uk`],
      subject: `Feedback about UCL Assistant`,
    })
    AnalyticsManager.logEvent(AnalyticsManager.events.SETTINGS_GIVE_FEEDBACK)
  }

  rateApp = async (): Promise<void> => {
    const isSupported = await StoreReview.isAvailableAsync()
    if (isSupported) {
      StoreReview.requestReview()
    }
    AnalyticsManager.logEvent(
      AnalyticsManager.events.SETTINGS_RATE_APP,
      { isSupported },
    )
  }

  renderDev = (): React.ReactNode => {
    const { user } = this.props
    return __DEV__ && (
      <View style={styles.section}>
        <HeaderText>UCL API Token</HeaderText>
        <Horizontal>
          <TextInput style={common.flex} value={user.token} />
          <SmallButton onPress={this.copyTokenToClipboard}>
            Copy
          </SmallButton>
        </Horizontal>
        <HeaderText>State</HeaderText>
        <BodyText>{JSON.stringify(user, null, 2)}</BodyText>
      </View>
    )
  }

  render(): React.ReactElement {
    const { user } = this.props
    return (
      <Page>
        <View style={styles.section}>
          <HeaderText>User</HeaderText>
          <BodyText>
            {`Logged in as ${user.fullName}`}
          </BodyText>
          <BodyText>
            {`Unique Person Identifier (UPI): ${user.upi}`}
          </BodyText>
          <Link
            onPress={this.signOut}
            style={styles.signOut}
            testID="sign-out-button"
          >
            Sign Out
          </Link>
          <View style={styles.settingView}>
            <BodyText
              style={styles.settingsText}
            >
              {
                user.settings.shouldTrackAnalytics
                  ? `Allow UCL Assistant to send analytics data to UCL API`
                  : `Disallow UCL Assistant from sending analytics data to `
                  + `UCL API`
              }
            </BodyText>
            <CheckBox
              testID="analytics-checkbox"
              style={styles.settingsCheckBox}
              isChecked={user.settings.shouldTrackAnalytics}
              onClick={this.toggleAnalytics}
            />
          </View>
        </View>
        <View style={styles.section}>
          <HeaderText>App Info</HeaderText>
          <Link href="https://uclapi.com">Website</Link>
          <Link
            onPress={this.navigateToFAQs}
            style={styles.faqButton}
            testID="faq-button"
          >
            Frequently Asked Questions
          </Link>
          <BodyText>
            {`Version: ${version}`}
          </BodyText>
          <Horizontal style={styles.releaseChannel}>
            {__DEV__ ? (
              <LiveIndicator>Developer Mode</LiveIndicator>
            ) : (
              <>
                  <BodyText>Release Channel: </BodyText>
                  <LiveIndicator>
                    {Constants.manifest.releaseChannel || `dev`}
                  </LiveIndicator>
              </>
            )}
          </Horizontal>
          <Link
            href={githubURL}
            testID="github-button"
          >
            Source Code
          </Link>
          <Link
            containerStyle={styles.feedbackButton}
            onPress={this.giveFeedback}
            testID="feedback-button"
          >
            Send Us Feedback
          </Link>
          {/* <Link
            onPress={this.rateApp}
          >
            Rate Us
          </Link> */}
        </View>
        <View style={styles.section}>
          <HeaderText>Credits</HeaderText>
          <BodyText>
            Created by Matt Bell (class of 2018) using the UCL API.
          </BodyText>
          <BodyText style={styles.textWithUpperMargin}>
            Currently managed by the UCL API Team: a group of
            students working together within ISD to improve UCL
            by building a platform on top of UCL&apos;s
            digital services for students.
          </BodyText>
          <BodyText style={styles.textWithUpperMargin}>
            Illustrations courtesy of the unDraw project,
            released under the MIT license.
          </BodyText>
        </View>
        {this.renderDev()}
      </Page>
    )
  }
}

const connector = connect(
  (state: AppStateType) => ({
    user: state.user,
  }),
  (dispatch: UserDispatch) => ({
    setShouldTrackAnalytics: (shouldTrackAnalytics) => dispatch(
      setShouldTrackAnalyticsAction(shouldTrackAnalytics),
    ),
    signOut: () => dispatch(signOutAction()),
  }),
)

type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(SettingsScreen)

/*

<View>
  <SubtitleText>Notifications</SubtitleText>
  <NotificationSwitch />
  <Button
    onPress={() => SettingsScreen.launchNotificationSettings()}
    style={styles.notificationSettingsButton}
  >
    Manage Notification Settings
  </Button>
</View>
<View style={styles.section}>
  <SubtitleText>Default Screen</SubtitleText>
  <BodyText>Set the default screen that shows when you open the app</BodyText>
  <Picker
    // selectedValue={this.state.language}
    // style={{height: 50, width: 100}}
    // onValueChange={(itemValue, itemIndex) =>
    //   this.setState({language: itemValue})
    // }
    >
    <Picker.Item label="Timetable" value="timetable" />
    <Picker.Item label="Study Spaces" value="studyspaces" />
  </Picker>
</View>

*/
