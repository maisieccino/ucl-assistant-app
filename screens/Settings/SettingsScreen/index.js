export { default, SettingsScreen } from "./SettingsScreen"
