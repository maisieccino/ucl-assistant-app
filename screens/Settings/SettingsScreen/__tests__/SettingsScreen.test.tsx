/**
 * @jest-environment jsdom
 */
import "react-native"

import React from 'react'
import { fireEvent, render } from "react-native-testing-library"

import { MailManager, WebBrowserManager } from '../../../../lib'
import * as packageJson from '../../../../package.json'
import { SettingsScreen } from "../SettingsScreen"

const {
  repository: {
    url: githubURL,
  },
} = packageJson

describe(`SettingsScreen`, () => {
  let wrapper
  const mockSignOut = jest.fn()
  const mockNavigate = jest.fn() as any
  const mockSetShouldTrackAnalytics = jest.fn()
  const mockProps = {
    navigation: {
      navigate: mockNavigate,
    } as any,
    setShouldTrackAnalytics: mockSetShouldTrackAnalytics,
    signOut: mockSignOut,
    user: {
      fullName: `Jeremy Bentham`,
      settings: {
        shouldTrackAnalytics: false,
      },
      token: `abc123`,
      upi: `upiupi11`,
    } as any,
  }

  const mockOpenLink = jest.fn(() => Promise.resolve())
  WebBrowserManager.openLink = mockOpenLink as any

  const mockComposeAsync = jest.fn(() => Promise.resolve())
  MailManager.composeAsync = mockComposeAsync

  beforeAll(() => {
    jest.useRealTimers()
  })

  beforeEach(() => {
    wrapper = render(<SettingsScreen {...mockProps} />)
  })

  it(`renders the SettingsScreen`, () => {
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it(`logs out when the signOut button is pressed`, () => {
    const { getByTestId } = wrapper
    fireEvent.press(getByTestId(`sign-out-button`))
    expect(mockSignOut).toHaveBeenCalledTimes(1)
  })

  it(`opens Github repository when Github button is pressed`, () => {
    const { getByTestId } = wrapper
    const githubButton = getByTestId(`github-button`)
    expect(githubButton.props.href === githubURL)
  })

  it(`opens FAQ page when the FAQ button is pressed`, () => {
    const { getByTestId } = wrapper
    const faqButton = getByTestId(`faq-button`)
    fireEvent.press(faqButton)

    expect(mockNavigate).toHaveBeenCalledTimes(1)
    expect(mockNavigate).toHaveBeenCalledWith(`FAQ`)
  })

  it(`opens mail client when the feedback button is pressed`, () => {
    const { getByTestId } = wrapper
    const feedbackButton = getByTestId(`feedback-button`)
    fireEvent.press(feedbackButton)

    expect(mockComposeAsync).toHaveBeenCalledTimes(1)
    expect(mockComposeAsync.mock.calls).toMatchSnapshot()
  })

  it(`dispatches action when analytics checkbox is toggled`, () => {
    const { getByTestId } = wrapper
    const analyticsCheckbox = getByTestId(`analytics-checkbox`)
    fireEvent(analyticsCheckbox, `click`)
    expect(mockSetShouldTrackAnalytics).toHaveBeenCalledTimes(1)
    expect(mockSetShouldTrackAnalytics).toHaveBeenCalledWith(true)
  })
})
