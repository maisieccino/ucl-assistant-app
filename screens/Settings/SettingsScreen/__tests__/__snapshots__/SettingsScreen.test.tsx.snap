// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsScreen opens mail client when the feedback button is pressed 1`] = `
Array [
  Array [
    Object {
      "body": "I've been using UCL Assistant and I just wanted to tell you ... 

Technical Information
{
  \\"deviceName\\": \\"Test Phone\\",
  \\"platform\\": {},
  \\"upi\\": \\"upiupi11\\"
}",
      "recipients": Array [
        "isd.apiteam@ucl.ac.uk",
      ],
      "subject": "Feedback about UCL Assistant",
    },
  ],
]
`;

exports[`SettingsScreen renders the SettingsScreen 1`] = `
Array [
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "flex": 0,
        },
        Object {
          "backgroundColor": "#fefefe",
        },
      ]
    }
  />,
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#fefefe",
          "flex": 1,
          "flexDirection": "column",
        },
        undefined,
      ]
    }
  >
    <View
      mainTabPage={true}
      onLayout={[Function]}
      style={
        Array [
          Array [
            Object {
              "backgroundColor": "#fefefe",
              "flex": 1,
              "flexDirection": "column",
            },
            undefined,
          ],
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <RCTScrollView
        contentContainerStyle={
          Array [
            Object {
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 10,
            },
            undefined,
          ]
        }
        keyboardDismissMode="on-drag"
        refreshControl={null}
        style={
          Array [
            Object {
              "backgroundColor": "#fefefe",
              "flex": 1,
              "paddingLeft": 20,
              "paddingRight": 20,
            },
            Object {
              "paddingLeft": 0,
              "paddingRight": 0,
            },
            undefined,
          ]
        }
      >
        <View>
          <View
            style={
              Object {
                "height": 10,
              }
            }
          />
          <View
            style={
              Object {
                "marginBottom": 15,
                "marginTop": 15,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu-light",
                    "fontSize": 28,
                    "marginBottom": 10,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              User
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 1,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              Logged in as Jeremy Bentham
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 1,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              Unique Person Identifier (UPI): upiupi11
            </Text>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#4D5061",
                      "fontFamily": "apercu",
                      "fontSize": 16,
                      "marginBottom": 1,
                      "marginTop": 1,
                    },
                    Object {
                      "color": "#2980b9",
                    },
                    Object {
                      "marginBottom": 10,
                      "marginTop": 10,
                    },
                  ]
                }
              >
                Sign Out
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#4D5061",
                      "fontFamily": "apercu",
                      "fontSize": 16,
                      "marginBottom": 1,
                      "marginTop": 1,
                    },
                    Object {
                      "flex": 0.9,
                    },
                  ]
                }
              >
                Disallow UCL Assistant from sending analytics data to UCL API
              </Text>
              <View
                accessible={true}
                focusable={true}
                isTVSelectable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "flex": 0.1,
                    "justifyContent": "center",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Image
                    source={1}
                    style={
                      Object {
                        "tintColor": undefined,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "marginBottom": 15,
                "marginTop": 15,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu-light",
                    "fontSize": 28,
                    "marginBottom": 10,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              App Info
            </Text>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#4D5061",
                      "fontFamily": "apercu",
                      "fontSize": 16,
                      "marginBottom": 1,
                      "marginTop": 1,
                    },
                    Object {
                      "color": "#2980b9",
                    },
                    Object {},
                  ]
                }
              >
                Website
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#4D5061",
                      "fontFamily": "apercu",
                      "fontSize": 16,
                      "marginBottom": 1,
                      "marginTop": 1,
                    },
                    Object {
                      "color": "#2980b9",
                    },
                    Object {
                      "marginBottom": 10,
                    },
                  ]
                }
              >
                Frequently Asked Questions
              </Text>
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 1,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              Version: 7.0.0
            </Text>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "marginBottom": 10,
                  },
                ]
              }
            >
              <ViewManagerAdapter_ExpoLinearGradient
                proxiedProperties={
                  Object {
                    "colors": Array [
                      4292290109,
                      4292309030,
                    ],
                    "endPoint": Array [
                      1,
                      0,
                    ],
                    "locations": undefined,
                    "startPoint": Array [
                      0,
                      1,
                    ],
                  }
                }
                style={
                  Array [
                    Object {
                      "borderRadius": 5,
                      "flex": 0,
                      "marginRight": 5,
                      "paddingHorizontal": 5,
                      "paddingVertical": 2,
                    },
                    Object {},
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#4D5061",
                        "fontFamily": "apercu",
                        "fontSize": 16,
                        "marginBottom": 1,
                        "marginTop": 1,
                      },
                      Object {
                        "color": "#fefefe",
                      },
                    ]
                  }
                >
                  Developer Mode
                </Text>
              </ViewManagerAdapter_ExpoLinearGradient>
            </View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#4D5061",
                      "fontFamily": "apercu",
                      "fontSize": 16,
                      "marginBottom": 1,
                      "marginTop": 1,
                    },
                    Object {
                      "color": "#2980b9",
                    },
                    Object {},
                  ]
                }
              >
                Source Code
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignSelf": "flex-start",
                  "marginVertical": 10,
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#4D5061",
                      "fontFamily": "apercu",
                      "fontSize": 16,
                      "marginBottom": 1,
                      "marginTop": 1,
                    },
                    Object {
                      "color": "#2980b9",
                    },
                    Object {},
                  ]
                }
              >
                Send Us Feedback
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "marginBottom": 15,
                "marginTop": 15,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu-light",
                    "fontSize": 28,
                    "marginBottom": 10,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              Credits
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 1,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              Created by Matt Bell (class of 2018) using the UCL API.
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 1,
                    "marginTop": 1,
                  },
                  Object {
                    "marginTop": 10,
                  },
                ]
              }
            >
              Currently managed by the UCL API Team: a group of students working together within ISD to improve UCL by building a platform on top of UCL's digital services for students.
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 1,
                    "marginTop": 1,
                  },
                  Object {
                    "marginTop": 10,
                  },
                ]
              }
            >
              Illustrations courtesy of the unDraw project, released under the MIT license.
            </Text>
          </View>
          <View
            style={
              Object {
                "marginBottom": 15,
                "marginTop": 15,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu-light",
                    "fontSize": 28,
                    "marginBottom": 10,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              UCL API Token
            </Text>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <TextInput
                allowFontScaling={true}
                placeholderTextColor="#6E707F"
                rejectResponderTermination={true}
                style={
                  Array [
                    Object {
                      "backgroundColor": "#eaeaea",
                      "borderColor": "#6E707F",
                      "borderRadius": 10,
                      "color": "#4D5061",
                      "fontFamily": "apercu",
                      "fontSize": 16,
                      "height": 50,
                      "marginBottom": 10,
                      "marginLeft": 2,
                      "marginRight": 2,
                      "marginTop": 10,
                      "minWidth": 200,
                      "paddingBottom": 5,
                      "paddingLeft": 10,
                      "paddingRight": 10,
                      "paddingTop": 5,
                    },
                    Object {
                      "flex": 1,
                    },
                  ]
                }
                underlineColorAndroid="transparent"
                value="abc123"
              />
              <View
                accessible={true}
                focusable={true}
                isTVSelectable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "opacity": 1,
                  }
                }
              >
                <ViewManagerAdapter_ExpoLinearGradient
                  proxiedProperties={
                    Object {
                      "colors": Array [
                        4279998859,
                        4283151357,
                      ],
                      "endPoint": Array [
                        1,
                        0,
                      ],
                      "locations": undefined,
                      "startPoint": Array [
                        0,
                        1,
                      ],
                    }
                  }
                  style={
                    Array [
                      Object {
                        "borderRadius": 8,
                        "elevation": 2,
                        "marginBottom": 5,
                        "marginTop": 5,
                        "minHeight": 55,
                        "minWidth": 150,
                        "padding": 8,
                        "paddingLeft": 16,
                        "paddingRight": 16,
                        "shadowColor": "#1B998B",
                        "shadowOffset": Object {
                          "height": 2,
                          "width": 0,
                        },
                        "shadowOpacity": 0.75,
                        "shadowRadius": 3,
                      },
                      Object {
                        "borderRadius": 8,
                        "elevation": 2,
                        "marginBottom": 5,
                        "marginTop": 5,
                        "minHeight": 30,
                        "minWidth": 90,
                        "padding": 8,
                        "paddingLeft": 16,
                        "paddingRight": 16,
                        "shadowColor": "#1B998B",
                        "shadowOffset": Object {
                          "height": 2,
                          "width": 0,
                        },
                        "shadowOpacity": 0.75,
                        "shadowRadius": 3,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#fefefe",
                          "flex": 1,
                          "fontFamily": "apercu",
                          "fontSize": 16,
                          "marginBottom": 1,
                          "marginTop": 1,
                          "textAlign": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    Copy
                  </Text>
                </ViewManagerAdapter_ExpoLinearGradient>
              </View>
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu-light",
                    "fontSize": 28,
                    "marginBottom": 10,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              State
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#4D5061",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 1,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              {
  "fullName": "Jeremy Bentham",
  "settings": {
    "shouldTrackAnalytics": false
  },
  "token": "abc123",
  "upi": "upiupi11"
}
            </Text>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </RCTSafeAreaView>,
]
`;
